File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    
    public static char[] cmds = {'U', 'D', 'R', 'L'};
    public static int[] dr = {0, 0, 1, -1};
    public static int[] dc = {1, -1, 0, 0};
    
    public int solution(String dirs) {
        
        char[] dir_arr = dirs.toCharArray();
        
        // 각 좌표에 0~121 사이의 번호를 부여한다.
        Set<Route> rs = new HashSet<>();
        int count = 0;
        int row, col;
        int nr, nc;
        int cur_point, next_point;
        
        row = 5;
        col = 5;
        
        for(char ch: dir_arr) {
            for(int i=0; i<4; i++) {
                
                if (ch == cmds[i]) {
                    nr = row + dr[i];
                    nc = col + dc[i];
                    
                    if (nr < 0 || nc < 0 || nr > 10 || nc > 10)
                        break;
                    
                    // 출발 좌표와 도착 좌표를 게산한다.
                    cur_point = (row*11) + col;
                    next_point = (nr*11) + nc;
                    Route tmp1 = new Route(cur_point, next_point);
                    Route tmp2 = new Route(next_point, cur_point);
                    if (!rs.contains(tmp1) && !rs.contains(tmp2)) {
                        rs.add(tmp1);
                        rs.add(tmp2);
                        count++;
                    }
                    
                    row = nr;
                    col = nc;
                    
                    break;
                    
                }
                
            }
            
        }
        
        int answer = count;
        return answer;
    }
    
    // 두 좌표의 번호가 같으면 같은 루트로 취급
    static class Route {
        public int src;
        public int dst;
        
        public Route(int src, int dst) {
            this.src = src;
            this.dst = dst;
        }
        
        @Override
        public boolean equals(Object o) {
            if (!(o instanceof Route))
                return false;
            
            Route r = (Route) o;
            if (this.src == r.src && this.dst == r.dst)
                return true;
           
            return false;
        }
        
        @Override
        public int hashCode() {
            return Objects.hash(src, dst);
        }
        
    }
    
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

char dir[4] = {'U', 'D', 'R', 'L'};
int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0, 0};

int solution(string dirs) {
    
    vector<vector<bool>> history(121, vector<bool>(4, false));
    
    // 각 좌표를 0~120의 번호로 표현한다.
    // 각 번호에서 U, D, R, L로 간 기록이 있는지 확인해서, 간 적이 없으면 카운트한다.
    int x = 0;
    int y = 0;
    int nx, ny;
    int node_idx, next_node_idx;
    int r_i;
    int count = 0;
    for(char ch: dirs) {
        
        for(int i=0; i<4; i++) {
            if (ch == dir[i]) {
                nx = x + dx[i];
                ny = y + dy[i];
                
                if (nx < -5 || ny < -5 || nx > 5 || ny > 5)
                    break;
                
                node_idx = (x+5) + ((5-y) * 11);
                next_node_idx = (nx+5) + ((5-ny) * 11);
                // U, R이면 인덱스+1로 D, L 만들기
                // 혹은 반대
                r_i = (i == 0 || i == 2) ? i+1 : i-1;
                
                if (history[node_idx][i] == false && history[next_node_idx][r_i] == false) {
                    count++;
                    history[node_idx][i] = true;
                    history[next_node_idx][r_i] = true;
                }
                
                x = nx;
                y = ny;
                
                break;
            }
        }
        
    }
    
    int answer = count;
    return answer;
}